### ISSUE TITLE
ffmpeg error – Invalid data found

### ISSUE THREAD

**Issue by gy-mate** [snobu/destreamer#110]
Hi! I've tried downloading an exact Microsoft Stream video three times, but I got this error for each try. I got this error previously for another Stream video, but I succeeded downloading that for the second try. What strange is, that – apart from these two Stream videos – I could download everything else flawlessly.

I'm not sure if this is an _ffmpeg_ or _destreamer_ error.

Below is the pasted code. I've redacted my username, the video's link and its title. If you need those, I can send them privately to you.

`MacBook:destreamer gymate1$ ./destreamer.sh -o /Users/gymate1/Downloads/ -u <username> -i <video link>`
`Using ffmpeg version 4.1 Copyright (c) 2000-2018 the FFmpeg developers`
`Username: <username>`
`Access token still good for 48 minutes.`
`Fetching metadata...`
`Downloading Video: <video title>`

`Spawning ffmpeg with access token and HLS URL. This may take a few seconds...
progress` `[░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░] 0% 0 0s`

`ffmpeg returned an error: https://euwe-1.api.microsoftstream.com/api/videos/manifest?api-version=1.0&playbackurl=https://cvcq2ehpzmewuhceldfzfzxrzh.streaming.mediaservices.windows.net/4ba149e0-7841-4c45-9478-7623ffe4704b/2-qjlxf7tmclcb4diahg6bz2o64e.ism/manifest(format=m3u8-aapl): Invalid data found when processing input`

`Unknown FFmpeg error`

`MacBook:destreamer gymate1$ `

**Comment by lukaarma** [comment]
I'm not really sure were the problem is in this one... because I downloaded the file that ffmpeg is choking on and it seems all good to me so..

Can you please try to run it with the verbose flag and post back here the result?

**Comment by gymate-1** [comment]
Oh, sorry, I should've done that.
Ran it with the verbose flag, but it worked now with this video. Here is the code anyway:

```
MacBook:destreamer gymate1$ ./destreamer.sh -o /Users/gymate1/Downloads/ -i <video link> -v
Using ffmpeg version 4.1 Copyright (c) 2000-2018 the FFmpeg developers

Video URLs:
[
  '<video link>'
]

Access token has expired.
Launching headless Chrome to perform the OpenID Connect dance...
Navigating to login page...
We are logged in.
Wrote access token to token cache.
At this point Chromium's job is done, shutting it down...

Video GUIDs:
[ '726a1c9a-80d2-4d3e-8a8d-ae98674a3403' ]
Fetching metadata...
Calling https://euwe-1.api.microsoftstream.com/api/videos/726a1c9a-80d2-4d3e-8a8d-ae98674a3403?api-version=1.4-private
Fresh access token dropped into .token_cache
[ '/Users/gymate1/Downloads/' ]

Downloading Video: <video title>

Spawning ffmpeg with access token and HLS URL. This may take a few seconds...
Using a fresh cookie.
progress [████████████████████████████████████████████████████████████] 100% 483.2kbits/s 0s

Download finished: /Users/gymate1/Downloads//<video name>.mp4
MacBook:destreamer gymate1$ 
```

But! I've tried it with another Stream video. First I got this error (not the one that I opened the issue for):

```
MacBook:destreamer gymate1$ ./destreamer.sh -o /Users/gymate1/Downloads/ -i <video link> -v
Using ffmpeg version 4.1 Copyright (c) 2000-2018 the FFmpeg developers

Video URLs:
[
  '<video link>'
]

Access token still good for 57 minutes.

Video GUIDs:
[ '0177c0a6-21b7-4891-91a9-3bdf13c399b9' ]
Fetching metadata...
Calling https://euwe-1.api.microsoftstream.com/api/videos/0177c0a6-21b7-4891-91a9-3bdf13c399b9?api-version=1.4-private
[ '/Users/gymate1/Downloads/' ]

Downloading Video: <video title>

Spawning ffmpeg with access token and HLS URL. This may take a few seconds...
Error: Request failed with status code 503
    at createError (/Users/gymate1/destreamer/node_modules/axios/lib/core/createError.js:16:15)
    at settle (/Users/gymate1/destreamer/node_modules/axios/lib/core/settle.js:17:12)
    at IncomingMessage.handleStreamEnd (/Users/gymate1/destreamer/node_modules/axios/lib/adapters/http.js:236:11)
    at IncomingMessage.emit (events.js:322:22)
    at endReadableNT (_stream_readable.js:1187:12)
    at processTicksAndRejections (internal/process/task_queues.js:84:21) {
  config: {
    url: 'https://euwe-1.api.microsoftstream.com/api/refreshtoken?api-version=1.4-private',
    method: 'get',
    headers: {
      Accept: 'application/json, text/plain, */*',
      Authorization: 'Bearer <token>',
      'User-Agent': 'axios/0.19.2'
    },
    transformRequest: [ [Function: transformRequest] ],
    transformResponse: [ [Function: transformResponse] ],
    timeout: 0,
    adapter: [Function: httpAdapter],
    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',
    maxContentLength: -1,
    validateStatus: [Function: validateStatus],
    data: undefined
  },
  request: ClientRequest {
    _events: [Object: null prototype] {
      socket: [Function],
      abort: [Function],
      aborted: [Function],
      error: [Function],
      timeout: [Function],
      prefinish: [Function: requestOnPrefinish]
    },
    _eventsCount: 6,
    _maxListeners: undefined,
    outputData: [],
    outputSize: 0,
    writable: true,
    _last: true,
    chunkedEncoding: false,
    shouldKeepAlive: false,
    useChunkedEncodingByDefault: false,
    sendDate: false,
    _removedConnection: false,
    _removedContLen: false,
    _removedTE: false,
    _contentLength: 0,
    _hasBody: true,
    _trailer: '',
    finished: true,
    _headerSent: true,
    socket: TLSSocket {
      _tlsOptions: [Object],
      _secureEstablished: true,
      _securePending: false,
      _newSessionPending: false,
      _controlReleased: true,
      _SNICallback: null,
      servername: 'euwe-1.api.microsoftstream.com',
      alpnProtocol: false,
      authorized: true,
      authorizationError: null,
      encrypted: true,
      _events: [Object: null prototype],
      _eventsCount: 10,
      connecting: false,
      _hadError: false,
      _parent: null,
      _host: 'euwe-1.api.microsoftstream.com',
      _readableState: [ReadableState],
      readable: true,
      _maxListeners: undefined,
      _writableState: [WritableState],
      writable: false,
      allowHalfOpen: false,
      _sockname: null,
      _pendingData: null,
      _pendingEncoding: '',
      server: undefined,
      _server: null,
      ssl: [TLSWrap],
      _requestCert: true,
      _rejectUnauthorized: true,
      parser: null,
      _httpMessage: [Circular],
      [Symbol(res)]: [TLSWrap],
      [Symbol(asyncId)]: 273,
      [Symbol(kHandle)]: [TLSWrap],
      [Symbol(kSetNoDelay)]: false,
      [Symbol(lastWriteQueueSize)]: 0,
      [Symbol(timeout)]: null,
      [Symbol(kBuffer)]: null,
      [Symbol(kBufferCb)]: null,
      [Symbol(kBufferGen)]: null,
      [Symbol(kCapture)]: false,
      [Symbol(kBytesRead)]: 0,
      [Symbol(kBytesWritten)]: 0,
      [Symbol(connect-options)]: [Object]
    },
    connection: TLSSocket {
      _tlsOptions: [Object],
      _secureEstablished: true,
      _securePending: false,
      _newSessionPending: false,
      _controlReleased: true,
      _SNICallback: null,
      servername: 'euwe-1.api.microsoftstream.com',
      alpnProtocol: false,
      authorized: true,
      authorizationError: null,
      encrypted: true,
      _events: [Object: null prototype],
      _eventsCount: 10,
      connecting: false,
      _hadError: false,
      _parent: null,
      _host: 'euwe-1.api.microsoftstream.com',
      _readableState: [ReadableState],
      readable: true,
      _maxListeners: undefined,
      _writableState: [WritableState],
      writable: false,
      allowHalfOpen: false,
      _sockname: null,
      _pendingData: null,
      _pendingEncoding: '',
      server: undefined,
      _server: null,
      ssl: [TLSWrap],
      _requestCert: true,
      _rejectUnauthorized: true,
      parser: null,
      _httpMessage: [Circular],
      [Symbol(res)]: [TLSWrap],
      [Symbol(asyncId)]: 273,
      [Symbol(kHandle)]: [TLSWrap],
      [Symbol(kSetNoDelay)]: false,
      [Symbol(lastWriteQueueSize)]: 0,
      [Symbol(timeout)]: null,
      [Symbol(kBuffer)]: null,
      [Symbol(kBufferCb)]: null,
      [Symbol(kBufferGen)]: null,
      [Symbol(kCapture)]: false,
      [Symbol(kBytesRead)]: 0,
      [Symbol(kBytesWritten)]: 0,
      [Symbol(connect-options)]: [Object]
    },
    _header: 'GET /api/refreshtoken?api-version=1.4-private HTTP/1.1\r\n' +
      'Accept: application/json, text/plain, */*\r\n' +
      'Authorization: Bearer <token>\r\n' +
      'User-Agent: axios/0.19.2\r\n' +
      'Host: euwe-1.api.microsoftstream.com\r\n' +
      'Connection: close\r\n' +
      '\r\n',
    _onPendingData: [Function: noopPendingOutput],
    agent: Agent {
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      defaultPort: 443,
      protocol: 'https:',
      options: [Object],
      requests: {},
      sockets: [Object],
      freeSockets: {},
      keepAliveMsecs: 1000,
      keepAlive: false,
      maxSockets: Infinity,
      maxFreeSockets: 256,
      maxCachedSessions: 100,
      _sessionCache: [Object],
      [Symbol(kCapture)]: false
    },
    socketPath: undefined,
    method: 'GET',
    insecureHTTPParser: undefined,
    path: '/api/refreshtoken?api-version=1.4-private',
    _ended: true,
    res: IncomingMessage {
      _readableState: [ReadableState],
      readable: false,
      _events: [Object: null prototype],
      _eventsCount: 3,
      _maxListeners: undefined,
      socket: [TLSSocket],
      connection: [TLSSocket],
      httpVersionMajor: 1,
      httpVersionMinor: 1,
      httpVersion: '1.1',
      complete: true,
      headers: [Object],
      rawHeaders: [Array],
      trailers: {},
      rawTrailers: [],
      aborted: false,
      upgrade: false,
      url: '',
      method: null,
      statusCode: 503,
      statusMessage: 'Service Unavailable',
      client: [TLSSocket],
      _consuming: false,
      _dumped: false,
      req: [Circular],
      responseUrl: 'https://euwe-1.api.microsoftstream.com/api/refreshtoken?api-version=1.4-private',
      redirects: [],
      [Symbol(kCapture)]: false
    },
    aborted: false,
    timeoutCb: null,
    upgradeOrConnect: false,
    parser: null,
    maxHeadersCount: null,
    reusedSocket: false,
    _redirectable: Writable {
      _writableState: [WritableState],
      writable: true,
      _events: [Object: null prototype],
      _eventsCount: 2,
      _maxListeners: undefined,
      _options: [Object],
      _redirectCount: 0,
      _redirects: [],
      _requestBodyLength: 0,
      _requestBodyBuffers: [],
      _onNativeResponse: [Function],
      _currentRequest: [Circular],
      _currentUrl: 'https://euwe-1.api.microsoftstream.com/api/refreshtoken?api-version=1.4-private',
      [Symbol(kCapture)]: false
    },
    [Symbol(kCapture)]: false,
    [Symbol(kNeedDrain)]: false,
    [Symbol(corked)]: 0,
    [Symbol(kOutHeaders)]: [Object: null prototype] {
      accept: [Array],
      authorization: [Array],
      'user-agent': [Array],
      host: [Array]
    }
  },
  response: {
    status: 503,
    statusText: 'Service Unavailable',
    headers: {
      'content-type': 'text/html; charset=us-ascii',
      server: 'Microsoft-HTTPAPI/2.0',
      date: 'Wed, 29 Apr 2020 21:38:36 GMT',
      connection: 'close',
      'content-length': '326'
    },
    config: {
      url: 'https://euwe-1.api.microsoftstream.com/api/refreshtoken?api-version=1.4-private',
      method: 'get',
      headers: [Object],
      transformRequest: [Array],
      transformResponse: [Array],
      timeout: 0,
      adapter: [Function: httpAdapter],
      xsrfCookieName: 'XSRF-TOKEN',
      xsrfHeaderName: 'X-XSRF-TOKEN',
      maxContentLength: -1,
      validateStatus: [Function: validateStatus],
      data: undefined
    },
    request: ClientRequest {
      _events: [Object: null prototype],
      _eventsCount: 6,
      _maxListeners: undefined,
      outputData: [],
      outputSize: 0,
      writable: true,
      _last: true,
      chunkedEncoding: false,
      shouldKeepAlive: false,
      useChunkedEncodingByDefault: false,
      sendDate: false,
      _removedConnection: false,
      _removedContLen: false,
      _removedTE: false,
      _contentLength: 0,
      _hasBody: true,
      _trailer: '',
      finished: true,
      _headerSent: true,
      socket: [TLSSocket],
      connection: [TLSSocket],
      _header: 'GET /api/refreshtoken?api-version=1.4-private HTTP/1.1\r\n' +
        'Accept: application/json, text/plain, */*\r\n' +
        'Authorization: Bearer <token>\r\n' +
        'User-Agent: axios/0.19.2\r\n' +
        'Host: euwe-1.api.microsoftstream.com\r\n' +
        'Connection: close\r\n' +
        '\r\n',
      _onPendingData: [Function: noopPendingOutput],
      agent: [Agent],
      socketPath: undefined,
      method: 'GET',
      insecureHTTPParser: undefined,
      path: '/api/refreshtoken?api-version=1.4-private',
      _ended: true,
      res: [IncomingMessage],
      aborted: false,
      timeoutCb: null,
      upgradeOrConnect: false,
      parser: null,
      maxHeadersCount: null,
      reusedSocket: false,
      _redirectable: [Writable],
      [Symbol(kCapture)]: false,
      [Symbol(kNeedDrain)]: false,
      [Symbol(corked)]: 0,
      [Symbol(kOutHeaders)]: [Object: null prototype]
    },
    data: '<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">\r\n' +
      '<HTML><HEAD><TITLE>Service Unavailable</TITLE>\r\n' +
      '<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>\r\n' +
      '<BODY><h2>Service Unavailable</h2>\r\n' +
      '<hr><p>HTTP Error 503. The service is unavailable.</p>\r\n' +
      '</BODY></HTML>\r\n'
  },
  isAxiosError: true,
  toJSON: [Function]
}


Unhandled error!
Timeout or fatal error, please check your downloads directory and try again
```

Then I tried it again, and got the error I opened the issue for:

```
MacBook:destreamer gymate1$ ./destreamer.sh -o /Users/gymate1/Downloads/ -i <video link> -v
Using ffmpeg version 4.1 Copyright (c) 2000-2018 the FFmpeg developers

Video URLs:
[
  '<video link>'
]

Access token still good for 40 minutes.

Video GUIDs:
[ '0177c0a6-21b7-4891-91a9-3bdf13c399b9' ]
Fetching metadata...
Calling https://euwe-1.api.microsoftstream.com/api/videos/0177c0a6-21b7-4891-91a9-3bdf13c399b9?api-version=1.4-private
[ '/Users/gymate1/Downloads/' ]

Downloading Video: <video title>


Spawning ffmpeg with access token and HLS URL. This may take a few seconds...
Using a fresh cookie.
progress [░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░] 0% 0 0s

ffmpeg returned an error: https://euwe-1.api.microsoftstream.com/api/videos/manifest?api-version=1.0&playbackurl=https://cvcq2ehpzmewuhceldfzfzxrzh.streaming.mediaservices.windows.net/bd9b044f-4ec1-4c28-8bb6-ad608e46cd12/1-pzcdprahcegckho26uomle4iza.ism/manifest(format=m3u8-aapl): Invalid data found when processing input



Unknown FFmpeg error 

MacBook:destreamer gymate1$ 
```

**Comment by lukaarma** [comment]
the 503 error you got is from our refreshToken function that has been giving us some problems and we are aware that it unstable/not behaving how it should and I think @snobu is already looking into it

The Issue that is the focus of this problem is one that I couldn't replicate and that I'm sure it's either ffmpeg crashing while waiting/parsing the response from the link we gave it or it's choking parsing the link itself, not sure.

I'm looking into a more stable way to call ffmpeg giving it an already downloaded playlist so we can be sure that's not a problem

Stay tuned for a possible update in the next days.
For now all I can suggest you is to brute force it and keep trying, sorry

**Comment by gymate-1** [comment]
Great, thank you so much guys! You're awesome! 

I've actually tried downloading the same video again and it worked this time. So I just have to try a few times.

**Comment by snobu** [comment]
The 503 is very random and i could not yet find a correlation with something we may be doing wrong. If you look closer it's a Service Unavailable response coming from some frontend (not even the API we're calling) - 
``
      statusCode: 503,
      statusMessage: 'Service Unavailable',
```

We could catch it and retry a few times, i'll look into it.

**Comment by snobu** [comment]
This should be fixed via https://github.com/snobu/destreamer/pull/130.
`git pull`, and rebuild to get the latest code.

**Comment by ali4heydari** [comment]
I have destreamer with latest commit `89a942eb24bb3a5774e69791dfb86c97c4c2740e` and get the ` Invalid data found when processing input` error.

node: 13.10.1
os: windows 10 x64

**Comment by s-monti97** [comment]
I fixed "Unhandled error! Timeout or fatal error, please check your downloads directory and try again" by just restarting my computer. Anyway, there is someway to increment token time? I'd like to download links inside a text file, but in 1 hour I just can get about 3 of them.

**Comment by snobu** [comment]
Yes. Wait for version 2.1, will get pushed next week. Pass -k to the CLI and tick Remember Me during login.

**Comment by s-monti97** [comment]
> Yes. Wait for version 2.1, will get pushed next week. Pass -k to the CLI and tick Remember Me during login.

That's wonderful, thanks!

**Comment by PARTHIBAN3691** [comment]
It is saying the token is invalid. I don't have any token. Previously it was working fine.
D:\DESTREAMER 64-BIT BY COMPUTING MENTOR>destreamer -i "https://web.microsoftstream.com/video/8a2598fb-c5ba-484b-a1bb-7e27a5ce3638"
Using ffmpeg version git-2020-05-01-39fb1e9 Copyright (c) 2000-2020 the FFmpeg developers

Error [InvalidTokenError]
    at Object.<anonymous> (C:\snapshot\destreamer\node_modules\jwt-decode\lib\index.js:9:31)
    at Module._compile (pkg/prelude/bootstrap.js:1320:22)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1165:10)
    at Module.load (internal/modules/cjs/loader.js:994:32)
    at Function.Module._load (internal/modules/cjs/loader.js:894:14)
    at Module.require (internal/modules/cjs/loader.js:1034:19)
    at Module.require (pkg/prelude/bootstrap.js:1225:31)
    at require (internal/modules/cjs/helpers.js:72:18)
    at Object.<anonymous> (C:\snapshot\destreamer\build\src\TokenCache.js:15:38)
    at Module._compile (pkg/prelude/bootstrap.js:1320:22) {
  message: 'Invalid token specified'
}


Unhandled error!
Timeout or fatal error, please check your downloads directory and try again

**Comment by snobu** [comment]
First make sure you're using the latest code by going a git clone in a new folder.
Follow the steps in the README carefully and it will work for you too.

**Comment by gy-mate** [comment]
I guess this was fixed in v2.1. Thank you very much, @snobu!
